### ISSUE TITLE
3th party authentication example

### ISSUE THREAD

**Issue by driesdroesbeke** [panva/node-oidc-provider#404]
I'm looking for an example where the user uses an external authentication form (idp,social,...) . 

If I can get it to work I would like to contribute to the examples folder with a social login example since a lot of users have been asking related questions.

****************************************
I still have a problem (re)using the same session when the user is redirected from the external IDP to my auth application.

### A) consumer app to oidc-provider
A user is redirected from a consumer app to http://myauthapp/identity/auth?client_id=foo&response_type=id_token&scope=openid&nonce=d

I changed the interactionUrl:
```js
interactionUrl: function interactionUrl(ctx, interaction) { // eslint-disable-line no-unused-vars
    return `/auth/login`;
},
```

### B) /auth/login (custom endpoint): show authentication options

this screen shows a list of 3the party authentication options and after a user clicks on a link the user is redirecten to an idp. (twitter(oauth), google(openid), idp(saml))

### C) authentication on external idp
user authenticates on idp and is redirected to http://myauthapp/auth/callback (custom endpoint)

### D) /auth/callback (custom endpoint)
I validate the saml and want to redirect the user to the originating app but I keep getting an SessionNotFound error ('interaction session id cookie not found') when i run
```js
return oidc.interactionFinished(req, res, results);
```
I've been playing around with *setProviderSession* but I did not find a proper solution to set/retrieve the session. 

How is this supposed to work?

**Comment by panva** [comment]
Another thing, your interaction root is /auth/login. So the interaction cookie is by default scoped to that path and anything “below” it. So by tge time the idp redirects to /auth/calback that cookie wont be found. 

Notice how the interaction example routes are all subresources of the interactionUrl result. 

I think that’s your issue.

**Comment by driesdroesbeke** [comment]
Thx! I didn't notice the path specific cookies.

Do you accept a pr with an example for this setup?

**Comment by ncphins** [comment]
I didn't really want to reopen an old issue, but since I believe others may be seeing this post and wondering as I did.

I have the same problem.  And I see that the path based cookies are the issue, but I am not sure how it is supposed to be solved.  I have the exact setup that was mentioned in the original post.

I authenticate to an external idp (SAML in this case) using passport which after authentication POSTs back to a callback.  I setup the callback to be "/interaction/callback" for my example.  The cookie is scoped to "/interaction/<uuid>" so when the callback happens, the cookie is not found.  So the session cannot be retrieved.  This is exactly as described above.

However, I am not sure how to solve it.  I cannot update the callback the idp uses to include the UUID.  It is registered with the provider and cannot be dynamic.  So I do not see how the cookie will ever be available.  As a result, I am not sure how I would be able to call interactionFinished.

If I manually re-write the cookie before I issue the authentication request and start the SAML authentication such that the path parameter is set to "/", the logic works fine and my user is authenticated and the code received successfully.  However I know that is not appropriate.  

I know I am doing something wrong or missing something.  Because I am sure this works.  But I am not sure what I am missing.  But since there were no followups, I believe the author of this thread has probably solved it.

Thanks for any help or insight!

**Comment by panva** [comment]
@ncphins if you can't think of a way to bounce the details to back a scoped path you may just as well eliminate the scoped cookies altogether.

```js
new Provider('...', {
  cookies: {
    short: {
      path: '/',
    },
  },
});
```

Note that not using scoped cookies means you'll be experiencing avoidable behaviour where two concurrent interactions will overwrite the cookie.

**Comment by panva** [comment]
This would also be handled for you if your interactions just didn't include the `uid` at all.

```js
new Provider('...', {
  interactionUrl() { return `/interaction`; },
});
```

**Comment by ncphins** [comment]
Ok.  Thanks.  I have it working without the scope, but the concurrent issue is real.  I will see if I can work out a more generic way with the providers I am using.  

I had it working modifying the interactionURL to not include the uuid.  But I felt like the concurrency problem would exist with that setup as well.

**Comment by panva** [comment]
Indeed, those two are equal in that regard. What I always used when third party idp was in play was an interaction session of my own. A completely separate mechanism, when the idp login was over i redirect to a scoped path which read the results from a cache and `interactionFinished` with those.

**Comment by panva** [comment]
You also don't NEED to use the helpers when you're out of scope, they're there to just make it easier.

As you start you idp login you can use e.g. the `state` parameter (i.e. have a it a signed JWT) to pass the uuid back to your callback route which isn't scoped, in that callback just mimic what [interactionResult](https://github.com/panva/node-oidc-provider/blob/v6.0.0-alpha.4/lib/provider.js#L188) does.

Something to this effect.
```js
const interaction = await provider.Interaction.find(uuid);
assert(interaction);
interaction.result = { ... };
await interaction.save(interaction.exp - epochTime());
req.redirect(interaction.returnTo);
```

**Comment by ncphins** [comment]
The state parameter is a really good idea.  Let me play around with that to see if I can make that work.  Thanks for all the help (and so quick).  I will let you know what I get working.

**Comment by Ineluki** [comment]
I'm having a similar issue so I'll ask this here, since I found the docs do not touch on this issue: In an app with a login outside of the provider and any interaction, how do I tell the provider that the user is already logged in when an interaction starts (or before?), so that the redirect to login never happens?
